feflow/protocols/nonequilibrium_cycling.py

```diff
@@ -165,26 +166,32 @@ def _execute(self, ctx, *, state_a, state_b, mapping, settings, **inputs):
         # Note: by default this is cached to ctx.shared/db.json so shouldn't
         # incur too large a cost
         self.logger.info("Parameterizing molecules")
-        small_mols_a = []
-        for comp in state_a.components.values():
-            if isinstance(comp, SmallMoleculeComponent):
-                small_mols_a.append(comp)
-
-        for comp in small_mols_a:
-            offmol = comp.to_openff()
-            system_generator.create_system(offmol.to_topology().to_openmm(),
-                                           molecules=[offmol])
-            if comp == ligand_a:
-                mol_b = ligand_b.to_openff()
-                system_generator.create_system(mol_b.to_topology().to_openmm(),
-                                               molecules=[mol_b])
+        # TODO: Refactor if/when gufe provides the functionality https://github.com/OpenFreeEnergy/gufe/issues/251
+        # The following creates a dictionary with all the small molecules in the states, with the structure:
+        #    Dict[SmallMoleculeComponent, openff.toolkit.Molecule]
+        state_a_small_mols = {component: component.to_openff() for component in state_a.components.values() if
+                              isinstance(component, SmallMoleculeComponent)}
+        state_b_small_mols = {component: component.to_openff() for component in state_b.components.values() if
```

Comment:
Am I correct in understanding here that "common" small molecules between the two states are effectively being duplicated here?

If so there are two considerations:

1. Given the complexity of some spectator molecules, this might add a bit of an overhead in duplicating partial charge generation. It's not the worst, but it's enough that if you have a multi-ring molecule you might find yourself idling for a little bit on what should be a short simulation.
2. Because we're seeing a lack of consistency between repeats of a partial charge generation (i.e. calling antechamber multiple times on the same inputs), there is a small risk that you could end up with non-transforming molecules that have different partial charges. This is a bit rarer, and eventually we'll have a fix, but it is still something that is likely to persist for a bit until we can fix things upstream.

Reply:
I see, that makes a lot of sense, and I agree. We probably want to avoid having to run into those issues. That explains why using the `Mapping` object here could be a better choice, just like you did on the [openfe side](https://github.com/OpenFreeEnergy/openfe/blob/5ddaad210ae9b41e5272ed9cd4e619f14541dac9/openfe/protocols/openmm_rfe/equil_rfe_methods.py#L680-L686). 

That approach works, but it seems a bit hard to read/understand and might be prone to errors in the future. Wouldn't using a `set` work here? Since the set operations should be okay to be used with the openff `Molecule` objects. For example as in:

```python
In [11]: mol_list = [benzene]*3 + [hexane]*4 + [cyclohexane]*7

In [12]: mol_list
Out[12]: 
[Molecule with name '' and SMILES '[H][c]1[c]([H])[c]([H])[c]([H])[c]([H])[c]1[H]',
 Molecule with name '' and SMILES '[H][c]1[c]([H])[c]([H])[c]([H])[c]([H])[c]1[H]',
 Molecule with name '' and SMILES '[H][c]1[c]([H])[c]([H])[c]([H])[c]([H])[c]1[H]',
 Molecule with name '' and SMILES '[H][C]([H])([H])[C]([H])([H])[C]([H])([H])[C]([H])([H])[C]([H])([H])[C]([H])([H])[H]',
 Molecule with name '' and SMILES '[H][C]([H])([H])[C]([H])([H])[C]([H])([H])[C]([H])([H])[C]([H])([H])[C]([H])([H])[H]',
 Molecule with name '' and SMILES '[H][C]([H])([H])[C]([H])([H])[C]([H])([H])[C]([H])([H])[C]([H])([H])[C]([H])([H])[H]',
 Molecule with name '' and SMILES '[H][C]([H])([H])[C]([H])([H])[C]([H])([H])[C]([H])([H])[C]([H])([H])[C]([H])([H])[H]',
 Molecule with name '' and SMILES '[H][C]1([H])[C]([H])([H])[C]([H])([H])[C]([H])([H])[C]([H])([H])[C]1([H])[H]',
 Molecule with name '' and SMILES '[H][C]1([H])[C]([H])([H])[C]([H])([H])[C]([H])([H])[C]([H])([H])[C]1([H])[H]',
 Molecule with name '' and SMILES '[H][C]1([H])[C]([H])([H])[C]([H])([H])[C]([H])([H])[C]([H])([H])[C]1([H])[H]',
 Molecule with name '' and SMILES '[H][C]1([H])[C]([H])([H])[C]([H])([H])[C]([H])([H])[C]([H])([H])[C]1([H])[H]',
 Molecule with name '' and SMILES '[H][C]1([H])[C]([H])([H])[C]([H])([H])[C]([H])([H])[C]([H])([H])[C]1([H])[H]',
 Molecule with name '' and SMILES '[H][C]1([H])[C]([H])([H])[C]([H])([H])[C]([H])([H])[C]([H])([H])[C]1([H])[H]',
 Molecule with name '' and SMILES '[H][C]1([H])[C]([H])([H])[C]([H])([H])[C]([H])([H])[C]([H])([H])[C]1([H])[H]']

In [13]: len(mol_list)
Out[13]: 14

In [14]: mol_set = set(mol_list)

In [15]: len(mol_set)
Out[15]: 3

```
That way, we can iterate through the set to parametrize the molecules and guarantee that they will only be passed once. I hope that makes sense. Thoughts?

Reply:
Oh I see the problem now, that doesn't work because there can be instances of the same molecule in different locations and/or with different conformers, and this would end up parametrizing only one of them. I think using the `Mapping` object is the way to go.